### PR TITLE
5141 database changes for item variants gaps

### LINKED ISSUE #5141: Database changes for Item Variants (GAPS)
## Is your feature request related to a problem? Please describe 👀

GAPS requires some item_variant features, completing this issue should implement the database changes needed for GAPS from this `item_variant` design.

There are several aspects of the GAPS work which require item variant capabilities
For reference this is the full [item variant PRD](https://docs.google.com/document/d/1fRZslRohRGaZnUo13a2eKZjfO0DGWwXtIFsbEah7A4Y/edit#heading=h.hyld495fuxe5)

We won't need all aspects of that; but we will need the ability to create and edit and sync item variants.

The variants will need to have these properties for the GAPS work:

| Name | Type   | Reference |
|:------ |:------ |:---- |
| Volume for different packaging levels | float | #3917 |
| Bundle items | reference to n item variants | #3923 |
| Temperature range | enum | #4368 |

See parent issue for more details...
https://github.com/msupply-foundation/open-msupply/issues/4789

## Describe the solution you'd like 🎁

2 new tables
```
    item_variant {
        TEXT id PK "NOT NULL"
        TEXT name "NOT NULL"
        TEXT item_link_id FK "REFERENCES item_link(id)"
        TEXT temperature_range "ENUM(+5, -20, -70)"  
        FLOAT doses_per_unit 
        TEXT manufacturer_id
        DATETIME deleted_datetime
    }
    packaging_variant {
        TEXT id PK "NOT NULL"
        TEXT name "NOT NULL"
        INT packaging_level "used to distingush tertiary, primary, etc"
        FLOAT pack_size
        FLOAT volume_per_unit
        DATETIME deleted_datetime
    }
```

Add `item_variant_id` but not `packaging_variant_id` (YET) to open-mSupply`stock_line`, and `invoice_line`.

A separate issue  will be created to add `oms_item_variant_id` to legacy `stock_line`, and `invoice_line`.

How might this look in an example?

image

```mermaid
erDiagram
    item_link {
        TEXT id PK "NOT NULL"
    }
    item_variant {
        TEXT id PK "NOT NULL"
        TEXT name "NOT NULL"
        TEXT item_link_id FK "REFERENCES item_link(id)"
        TEXT temperature_range "ENUM(+5, -20, -70)"  
        FLOAT doses_per_unit 
        TEXT manufacturer_id
        DATETIME deleted_datetime
    }
    packaging_variant {
        TEXT id PK "NOT NULL"
        TEXT name "NOT NULL"
        TEXT item_variant_id "REFERENCES item_variant(id) FUTURE"
        TEXT internal_packaging_variant_id FK "REFERENCES packaging_variant(id) FUTURE"
        INT pack_type "Box, Carton, Etc FUTURE"
        INT packaging_level "used to distingush tertiary, primary, etc"
        FLOAT pack_size
        FLOAT volume_per_unit
        DATETIME deleted_datetime
    }
    barcode {
        TEXT id FK "NOT NULL"
        TEXT item_link_id FK "REFERENCES item_link(id)"
        TEXT item_variant_id FK "REFERENCES item_variant(id)"
        TEXT packaging_variant FK "REFERENCES packaging_variant(id) FUTURE"
        TEXT barcode "UNIQUE"
        TEXT universal_code 
    }
    stock_line {
        TEXT id PK "NOT NULL"
        TEXT item_variant_id FK "REFERENCES item_variant(id)"
        TEXT packaging_variant FK "REFERENCES packaging_variant(id) FUTURE"
    }
    invoice_line {
        TEXT id PK "NOT NULL"
        TEXT item_link_id FK "REFERENCES item_link(id)"
        TEXT item_variant_id FK "REFERENCES item_variant(id)"
        TEXT packaging_variant FK "REFERENCES packaging_variant(id) FUTURE"
    }
item_link ||--o{ item_variant : ""
item_variant ||--o{ packaging_variant :""
packaging_variant ||--o{ packaging_variant : "internal_variant"
barcode ||--o{ item_variant : ""
barcode ||--o{ item_link :""
stock_line ||--o{ item_link : ""
stock_line ||--o{ item_variant : ""
invoice_line ||--o{ item_link : ""
invoice_line ||--o{ item_variant : ""
```

### Describe alternatives you've considered 💭

** Single Item Variants Table with packaging details **

After taking on board various feedback, and trying to design for future goals not just GAPS.
This issue suggests a item_variant table closer to the original design in the PRD.

**Item Variant**
```
        TEXT id PK "NOT NULL"
        TEXT name "NOT NULL"
        TEXT item_link_id FK "REFERENCES item_link(id)"
        TEXT internal_item_variant_id FK "REFERENCES item_variant(id) NOT NEEDED FOR GAPS"
        TEXT temperature_range "ENUM(+5, -20, -70)"
        INT packaging_level "used to distinguish tertiary, primary, etc"
        INT pack_type "Box, Carton, Etc NOT NEEDED FOR GAPS"
        INT num_units
        FLOAT doses_per_unit 
        FLOAT volume_per_pack
        FLOAT weight_per_pack "NOT NEEDED FOR GAPS" 
        TEXT manufacturer_id
```

How might this look in an example?

image

NOTE: In the gaps master data we have 7 different level 1 item_variants for BCG, once you include the secondard and tertiary packaging there'll be 21 rows of item variants for a BCG vaccine, which is a consideration I think...

You can see that doses_per_unit needs to be duplicated on every row but otherwise it looks relatively clean to me.

One issue I found creating this example is that existing GAPS master data has a volume per unit stored against secondary, tertiary etc. I might not have looked hard enough though...

My design has a volume per pack, as I was assuming that would be easier to get and deal with as a end user.
However maybe we store a `volume_per_unit` but allow users to enter the packsize and total volume to calculate this they want?

```mermaid
erDiagram
    item_link {
        TEXT id PK "NOT NULL"
    }
    item_variant {
        TEXT id PK "NOT NULL"
        TEXT name "NOT NULL"
        TEXT item_link_id FK "REFERENCES item_link(id)"
        TEXT internal_item_variant_id FK "REFERENCES item_variant(id) NOT NEEDED FOR GAPS"
        TEXT temperature_range "ENUM(+5, -20, -70)"
        INT packaging_level "used to distingush tertiary, primary, etc"
        INT pack_type "Box, Carton, Etc NOT NEEDED FOR GAPS"
        INT num_units
        FLOAT doses_per_unit 
        FLOAT volume_per_unit
        TEXT manufacturer_id

    }
    barcode {
        TEXT id FK "NOT NULL"
        TEXT item_link_id FK "REFERENCES item_link(id)"
        TEXT item_variant_id FK "REFERENCES item_variant(id)"
        TEXT barcode "UNIQUE"
        TEXT universal_code 
    }
    stock_line {
        TEXT id PK "NOT NULL"
        TEXT item_variant_id FK "REFERENCES item_variant(id)"
    }
    invoice_line {
        TEXT id PK "NOT NULL"
        TEXT item_link_id FK "REFERENCES item_link(id)"
        TEXT item_variant_id FK "REFERENCES item_variant(id)"
    }
item_link ||--o{ item_variant : ""
item_variant ||--o{ item_variant : "internal_variant"
barcode ||--o{ item_variant : ""
barcode ||--o{ item_link :""
stock_line ||--o{ item_link : ""
stock_line ||--o{ item_variant : ""
invoice_line ||--o{ item_link : ""
invoice_line ||--o{ item_variant : ""
```

**Add temperature ranges rather than enum?**
An enum is easy for gaps work, e.g. `-20` is separated out, but to make the system more flexible, and to potentially allow for more specific ranges for item variants we could replace it with `min_temperature`, `max_temperature` or potentially have both?

**Add code for item variants?**
Looks like there could be a number of variants for an item, would a code be helpful rather than a big long name?

**Store item variants in the barcode table**
I think logically this may well  make sense in the future, when all manufacturers have barcodes on items, and we have good support for 
However here's my arguments for not doing that now.

1. Having records in a barcode table with no barcode seems nonsensical!
2. Renaming barcode table to item_variant would have hard to understand consequences 
3. There's a lot of changes we'd like to make to barcode table which would be more difficult to manage with 4d & rust in the picture
4. There's potential for both a mSupply assigned and manufacturer assigned barcode for the same item. So having a 1 - many for this slight chance feels potentially helpful?

**Separate out packaging variants from item variant**
This was essentially allowing us to have one `item_variant` for say Pfizer Covid-19 Vaccine, which would then have different packaging options like `Container`, `Crate`, `Box`, `Tray`, etc.
I think the the addition of the `internal_variant` in the design we can achieve a similar goal without having the additional table involved adding additional complex it.
By adding the packaging_level to item variant, you can still query the table to find appropriate options at the facility level if defined (e.g. for gaps)

**item_variant proposed in original issue**
This records things like `volume_per_secondary_packaging` against an item variant.
While it might serve for gaps, doesn't cover off the use cases for item_variant in the [PRD](https://docs.google.com/document/d/1fRZslRohRGaZnUo13a2eKZjfO0DGWwXtIFsbEah7A4Y/edit#heading=h.hyld495fuxe5)
Such as being able to receipt into a store `45 Cartons` and the system would calculate the number of units etc automatically.

I think a little bit more work on now will pay off in the future!

### Additional context 💌

`num_units` is essentially `pack_size` I think?
I feel like this is more descriptive, especially if you're thinking of how many units is in a container or something, `pack_size` would be easier to miss-understand, like its it the how many boxes are in the carton or how many units are there?
Potentially `pack_size` would be correct for our environment though?

### Moneyworks Jobcode 🧰

CCEIDEV

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #5141 

# 👩🏻‍💻 What does this PR do?

Adds the `item_variant` and `packaging_variant` tables.
`item_variant_row` and `packaging_variant_row` repos.

Query, Service,Layer etc in a later pr...

## 💌 Any notes for the reviewer?

# 🧪 Testing

- [ ] No real testing required yet.

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [X] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

File: server/repository/src/db_diesel/item_variant/item_variant_row.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+#[derive(
+    Clone, Queryable, Insertable, AsChangeset, Debug, PartialEq, Default, Serialize, Deserialize,
+)]
+#[diesel(table_name = item_variant)]
+pub struct ItemVariantRow {
+    pub id: String,
+    pub name: String,
+    pub item_link_id: String,
+    pub doses_per_unit: Option<f64>,
+    pub manufacturer_link_id: Option<String>,
```
Comment (jmbrunskill) [inline server/repository/src/db_diesel/item_variant/item_variant_row.rs:31]: `cold_storage_type` to come soon...

File: server/repository/src/migrations/v2_04_00/item_variant.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    fn migrate(&self, connection: &StorageConnection) -> anyhow::Result<()> {
+        // Item variant Table
+        sql!(
+            connection,
+            r#"
+            CREATE TABLE item_variant (
+                id TEXT PRIMARY KEY NOT NULL,
+                name TEXT NOT NULL,
+                item_link_id TEXT NOT NULL REFERENCES item_link(id),
+                --TODO temperature_range_id TEXT REFERENCES temperature_range(id),
```
Comment (lache-melvin) [inline server/repository/src/migrations/v2_04_00/item_variant.rs:20]: This actually raises a little question for me - is your thinking to add the column here and reset DB seeing as we're in dev, vs an `alter table add column` in another fragment? Only do that once we're in RC? I had been just always creating migrations, but it's an interesting balance (dev experience during the release, vs reading back through migrations and seeing the related columns all together 👀)

**Review by lache-melvin — APPROVED** [review]
Looks good to me! Thanks James 🙏

File: server/repository/src/migrations/v2_04_00/item_variant.rs (hunk 2 above)
Comment (CarlosNZ) [reply server/repository/src/migrations/v2_04_00/item_variant.rs:20]: I think he's just waiting for me, in which case we'd probably update this PR before merge? Anyway, I'll publish the Repository layer of mine shortly.

Comment (jmbrunskill) [reply server/repository/src/migrations/v2_04_00/item_variant.rs]: Yup, rebased now so should be all good once both are merged!

Feel free to merge without me if you want this in Develop!

**Pull request merged**